src/key/key.js

```diff
@@ -360,47 +359,77 @@ class Key {
+      ).catch(() => {}); // invalid signatures are discarded, to avoid breaking the key
+
+      if (directSignature && helper.isDataExpired(primaryKey, directSignature, date)) {
+        throw new Error('Primary key is expired through a direct signature');
+      }
+    }
   }
 
   /**
    * Returns the latest date when the key can be used for encrypting, signing, or both, depending on the `capabilities` paramater.
    * When `capabilities` is null, defaults to returning the expiry date of the primary key.
    * Returns null if `capabilities` is passed and the key does not have the specified capabilities or is revoked or invalid.
    * Returns Infinity if the key doesn't expire.
-   * @param  {encrypt|sign|encrypt_sign} capabilities, optional
-   * @param  {module:type/keyid~KeyID} keyID, optional
-   * @param  {Object} userID, optional user ID
-   * @param {Object} [config] - Full configuration, defaults to openpgp.config
+   * @param  {encrypt|sign|encrypt_sign} [capabilities] - capabilities to look up
+   * @param  {module:type/keyid~KeyID} [keyID] - key ID of the specific key to check
+   * @param  {Object} [userID] - User ID to consider instead of the primary user
+   * @param  {Object} [config] - Full configuration, defaults to openpgp.config
    * @returns {Promise<Date | Infinity | null>}
    * @async
    */
   async getExpirationTime(capabilities, keyID, userID, config = defaultConfig) {
-    const primaryUser = await this.getPrimaryUser(null, userID, config);
-    const selfCert = primaryUser.selfCertification;
-    const keyExpiry = helper.getExpirationTime(this.keyPacket, selfCert);
-    const sigExpiry = selfCert.getExpirationTime();
-    let expiry = keyExpiry < sigExpiry ? keyExpiry : sigExpiry;
+    const toDateOrInfinity = value => (value === Infinity ? value : new Date(value));
+    const minNotNull = (...args) => { // Like Math.min but aware of `null`, to avoid treating them as 0
+      const notNull = args.filter(arg => arg !== null);
+      if (notNull.length === 0) return null;
+      return Math.min(...notNull);
+    };
+
+    let primaryKeyExpiry;
+    try {
+      const { selfCertification } = await this.getPrimaryUser(null, userID, config);
+      const keyExpiry = helper.getKeyExpirationTime(this.keyPacket, selfCertification);
+      const sigExpiry = selfCertification.getExpirationTime();
+      // TODO check direct signatures
+      primaryKeyExpiry = toDateOrInfinity(Math.min(keyExpiry, sigExpiry));
+    } catch (e) {
+      primaryKeyExpiry = null;
+    }
+    if (!capabilities) return primaryKeyExpiry;
+
+    // loop through subkeys based on requested capabilities and retrieve their expiration times
+    const getLatestKeyExpiry = async keys => (keys.length === 0 ? null : toDateOrInfinity(
+      Math.max(...await Promise.all(
+        keys.map(async key => (key instanceof SubKey ? key.getExpirationTime(null, config) : primaryKeyExpiry))
```

Comment:
You don't need this `async` here, as there's no `await`

Reply:
Promise.all is awaited

Reply:
I mean the one on line 410, not the one on line 408